### PR TITLE
feat: support gleam lang

### PR DESCRIPTION
**Pull request by cdaringe**
## Changes

Add gleam-lang manager support

- [x] figure out what to do for getting `manifest.toml` lockfile updated
- [x] tests
- [x] docs? need to figure out where to put docs

## Context

see #26744

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

Run in debug mode! Sending PRs over to the poorly named https://github.com/cdaringe/renovate-npm-testerooni/pull/3

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gleam/artifacts.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import is from '@sindresorhus/is';
```
Comment (cdaringe) [inline lib/modules/manager/gleam/artifacts.ts:1]: ignore this! copied from bun, was hoping to adapt the manifest updater using a gleam docker image. i don't think we have a gleam docker image publishing yet? i see one from a few years ago

File: lib/modules/manager/gleam/extract.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import assert from 'assert';
+// import semver, { satisfies } from 'semver';
```
Comment (cdaringe) [inline lib/modules/manager/gleam/extract.ts:2]:
```suggestion
```

File: lib/modules/manager/gleam/index.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import type { Category } from '../../../constants';
+import { GithubTagsDatasource } from '../../datasource/github-tags';
+import * as version from '../../versioning/hex';
+
+// export { extractPackageFile } from './extract';
+
```
Comment (cdaringe) [inline lib/modules/manager/gleam/index.ts:6]:
```suggestion
```

**Comment by rarkins** [comment]
Is a new binary needed for updating lock files? If so it will need to be added to GitHub.com/containerbase/base

File: lib/constants/category.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@ const Categories = [
   'dotnet',
   'elixir',
   'golang',
+  'gleam',
```
Comment (rarkins) [inline lib/constants/category.ts:14]: Not needed  if it's a category of one

File: lib/modules/manager/gleam/artifacts.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    if (isLockFileMaintenance) {
+      await deleteLocalFile(lockFileName);
+    }
+
+    const execOptions: ExecOptions = {
+      cwdFile: packageFileName,
+      docker: {},
+      toolConstraints: [
+        {
+          toolName: 'gleam',
```
Comment (rarkins) [inline lib/modules/manager/gleam/artifacts.ts:55]: Needs adding containerbase/base repo

File: lib/modules/manager/gleam/index.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import * as version from '../../versioning/hex';
+
+export const displayName = 'gleam';
+export const url = 'https://gleam.run/';
+
+export { extractPackageFile } from './extract';
+export { updateArtifacts } from './artifacts';
+
+export const defaultConfig = {
+  fileMatch: ['^gleam.toml$'],
```
Comment (rarkins) [inline lib/modules/manager/gleam/index.ts:12]: Only in root did supported?

File: lib/modules/manager/gleam/index.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import type { Category } from '../../../constants';
+import { GithubTagsDatasource } from '../../datasource/github-tags';
+import * as version from '../../versioning/hex';
```
Comment (rarkins) [inline lib/modules/manager/gleam/index.ts:3]:
```suggestion
import * as hexVersioning from '../../versioning/hex';
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gleam/index.ts (hunk 6 above)
Comment (cdaringe) [reply lib/modules/manager/gleam/index.ts:12]: yep

File: lib/constants/category.ts (hunk 4 above)
Comment (cdaringe) [reply lib/constants/category.ts]: 441743b7a

File: lib/modules/manager/gleam/artifacts.ts (hunk 5 above)
Comment (cdaringe) [reply lib/modules/manager/gleam/artifacts.ts:52]: huh. in local it ran great and fast... guessing it didnt actually use containers 😄 .

will drop a cross ref when ready

File: lib/modules/manager/gleam/artifacts.spec.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+jest.mock('fs-extra');
+
+const exec = mocked(_exec);
+const fs = mocked(_fs);
+
+const globalConfig: RepoGlobalConfig = {
+  localDir: '',
+};
+
+describe('modules/manager/gleam/artifacts', () => {
```
Comment (cdaringe) [inline lib/modules/manager/gleam/artifacts.spec.ts:21]: this is depressingly large amount of copy paste from bun... but it's good and all relevant!

File: lib/modules/manager/gleam/artifacts.ts (hunk 5 above)
Comment (cdaringe) [reply lib/modules/manager/gleam/artifacts.ts:52]: https://github.com/containerbase/base/pull/2018

File: lib/modules/manager/gleam/__fixtures__/gleam.dev_deps_only.toml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+name = "test_gleam_toml"
```
Comment (viceice) [inline lib/modules/manager/gleam/__fixtures__/gleam.dev_deps_only.toml:1]: inline those small fixtures to the spec test file and use `codeBlock` string template helper

File: lib/modules/manager/gleam/__fixtures__/gleam.prod_and_dev_deps.toml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+name = "test_gleam_toml"
```
Comment (viceice) [inline lib/modules/manager/gleam/__fixtures__/gleam.prod_and_dev_deps.toml:1]: ☝️

File: lib/modules/manager/gleam/artifacts.spec.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+import { GlobalConfig } from '../../../config/global';
+import type { RepoGlobalConfig } from '../../../config/types';
+import { TEMPORARY_ERROR } from '../../../constants/error-messages';
+import { exec as _exec } from '../../../util/exec';
+import { ExecError } from '../../../util/exec/exec-error';
+import type { UpdateArtifact } from '../types';
+import { updateArtifacts } from './artifacts';
+
+jest.mock('../../../util/exec');
+jest.mock('fs-extra');
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.spec.ts:12]: no, mock `util/fs` instead

File: lib/modules/manager/gleam/artifacts.spec.ts [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+      ];
+      expect(await updateArtifacts(updateArtifact)).toBeNull();
+    });
+
+    it('returns null if lock content unchanged', async () => {
+      updateArtifact.updatedDeps = [
+        { manager: 'gleam', lockFiles: ['manifest.toml'] },
+      ];
+      const oldLock = Buffer.from('old');
+      fs.readFile.mockResolvedValueOnce(oldLock as never);
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.spec.ts:56]: you don't need this cast when using util/fs

File: lib/modules/manager/gleam/artifacts.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+
+  if (is.emptyArray(updatedDeps) && !isLockFileMaintenance) {
+    logger.debug('No updated gleam deps - returning null');
+    return null;
+  }
+
+  // Find the first gleam dependency in order to handle mixed manager updates
+  const lockFileName = updatedDeps.find((dep) => dep.manager === 'gleam')
+    ?.lockFiles?.[0];
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.ts:28]: isn't the lockfile always in repo root?  you should simply use the siblings fs helper to find it relative to the package file.

File: lib/modules/manager/gleam/artifacts.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+      await deleteLocalFile(lockFileName);
+    }
+
+    const execOptions: ExecOptions = {
+      cwdFile: packageFileName,
+      docker: {},
+      toolConstraints: [
+        {
+          toolName: 'gleam',
+          constraint: updateArtifact?.config?.constraints?.gleam,
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.ts:53]:
```suggestion
          constraint: config.constraints?.gleam,
```

File: lib/modules/manager/gleam/artifacts.ts [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        },
+      ],
+    };
+
+    await exec('gleam deps download', execOptions);
+    const newLockFileContent = await readLocalFile(lockFileName);
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.ts:59]: is the lockfile a binary or text file?

File: lib/modules/manager/gleam/artifacts.spec.ts [hunk 16]
```diff
@@ -0,0 +1,118 @@
+import _fs from 'fs-extra';
+import { mocked } from '../../../../test/util';
+import { GlobalConfig } from '../../../config/global';
+import type { RepoGlobalConfig } from '../../../config/types';
+import { TEMPORARY_ERROR } from '../../../constants/error-messages';
+import { exec as _exec } from '../../../util/exec';
+import { ExecError } from '../../../util/exec/exec-error';
+import type { UpdateArtifact } from '../types';
+import { updateArtifacts } from './artifacts';
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.spec.ts:9]:
```suggestion
import { updateArtifacts } from '.';
```

File: lib/modules/manager/gleam/extract.spec.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+import * as gleamManager from '.';
+
+describe('modules/manager/gleam/extract', () => {
+  it('should extract dev and prod dependencies', () => {
+    const gleamTomlString = Fixtures.get('gleam.prod_and_dev_deps.toml');
+    const extracted = gleamManager.extractPackageFile(
+      gleamTomlString,
+      'gleam.toml',
+    );
+    expect(extracted).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.spec.ts:11]: no snapshots, read our best practices develop guide

File: lib/modules/manager/gleam/extract.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
+import assert from 'assert';
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:1]: wrong, remove
```suggestion
```

File: lib/modules/manager/gleam/extract.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+type GleamToml = {
+  name: string;
+  dependencies?: Record<string, string>;
+  ['dev-dependencies']?: Record<string, string>;
+};
+
+function parseGleamToml(gleamTomlString: string): GleamToml {
+  const toml = parseToml(gleamTomlString) as GleamToml;
+  assert(toml, 'invalid toml');
+  assert(typeof toml.name === 'string');
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:14]: use zod schema validation instead

File: lib/modules/manager/gleam/index.ts [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+export { extractPackageFile } from './extract';
+export { updateArtifacts } from './artifacts';
+
+export const defaultConfig = {
+  fileMatch: ['^gleam.toml$'],
+  versioning: hexVersioning.id,
+};
+
+export const supportsLockFileMaintenance = true;
+export const supportedDatasources = [GithubTagsDatasource.id];
```
Comment (viceice) [inline lib/modules/manager/gleam/index.ts:16]:
```suggestion
export const supportedDatasources = [HexDatasource.id];
```

File: lib/modules/manager/gleam/extract.ts [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+  dev,
+}: {
+  name: string;
+  version: string;
+  dev: boolean;
+}): PackageDependency {
+  return {
+    depName: name,
+    depType: dev ? 'devDependencies' : 'dependencies',
+    datasource: 'hex',
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:30]:
```suggestion
    datasource: HexDatasource.id,
```

File: lib/modules/manager/gleam/extract.ts [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+export function extractPackageFile(
+  content: string,
+  packageFile?: string,
+): PackageFileContent | null {
+  const deps: PackageDependency[] =
+    packageFile === 'gleam.toml'
+      ? extractGleamTomlDeps(parseGleamToml(content))
+      : [];
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:53]:
```suggestion
  const deps = extractGleamTomlDeps(parseGleamToml(content));
```
simplify

File: lib/modules/manager/gleam/extract.ts [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+  return { deps, lockFiles: ['manifest.toml'] };
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:54]:
```suggestion
  return deps.length ? { deps, lockFiles: ['manifest.toml'] } : null;
```

File: lib/modules/manager/gleam/index.ts [hunk 24]
```diff
@@ -0,0 +1,16 @@
+import { GithubTagsDatasource } from '../../datasource/github-tags';
```
Comment (viceice) [inline lib/modules/manager/gleam/index.ts:1]:
```suggestion
import { HexDatasource } from '../../datasource/hex';
```

File: lib/modules/manager/gleam/extract.ts [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+    datasource: 'hex',
+    currentValue: version,
+  };
+}
+
+function extractGleamTomlDeps(gleamToml: GleamToml): PackageDependency[] {
+  return [
+    ...Object.entries(gleamToml.dependencies ?? {}).map(([name, version]) =>
+      toPackageDep({ name, version, dev: false }),
+    ),
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:39]: extract function and pass toml key to simplify things

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gleam/artifacts.ts (hunk 15 above)
Comment (cdaringe) [reply lib/modules/manager/gleam/artifacts.ts:59]: text. i'm adding the encoding param now

File: lib/modules/manager/gleam/artifacts.ts (hunk 13 above)
Comment (cdaringe) [reply lib/modules/manager/gleam/artifacts.ts:28]: sure. there is existing modeling for lockFiles in the `PackageDep` type, and I observed other modules using this. I have done what you've requested, but IMHO continuation of the `lockFiles` seems more consistent with other pkg managers? for now, new commits coming in refactoring to use getSiblingFile

File: lib/modules/manager/gleam/readme.md [hunk 26]
```diff
@@ -0,0 +1,3 @@
+# gleam-manager
+
+Updates `gleam.toml` with latest and greatest dependencies.
```
Comment (cdaringe) [inline lib/modules/manager/gleam/readme.md:3]:
```suggestion
Updates `gleam.toml` and/or `manifest.toml` with latest and greatest dependencies.
```

**Comment by cdaringe** [comment]
Lockfile maintenance also lookin good: https://github.com/cdaringe/gserde/pull/8

**Comment by cdaringe** [comment]
My goodness, I thought i was waiting for reviews 🤦 . Builds were failing, my bad! Hopefully these go green now that tests+coverage _should_ be fixed.

File: lib/modules/manager/gleam/extract.ts [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+    ...toPackageDeps({
+      deps: gleamToml['dev-dependencies'],
+      dev: true,
+    }),
+  ];
+}
+
+export function extractPackageFile(content: string): PackageFileContent | null {
+  const deps = extractGleamTomlDeps(GleamToml.parse(content));
+  /* istanbul ignore next reason: static analysis sufficient */
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:46]: ? nope, add a test

File: lib/modules/manager/gleam/extract.ts [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+  version,
+  dev,
+}: {
+  name: string;
+  version: string;
+  dev?: boolean;
+}): PackageDependency {
+  return {
+    depName: name,
+    depType: dev ? 'devDependencies' : 'dependencies',
```
Comment (viceice) [inline lib/modules/manager/gleam/extract.ts:16]: we usually use the section key for this. that will make it more easily extendable in future

File: lib/modules/manager/gleam/artifacts.ts [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+      {
+        file: {
+          type: 'addition',
+          path: lockFileName,
+          contents: newLockFileContent,
+        },
+      },
+    ];
+  } catch (err) {
+    /* istanbul ignore next reason: static analysis sufficient */
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.ts:74]: same, add a test and simply throw appropriate error from read file

File: lib/modules/manager/gleam/artifacts.ts [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+        {
+          toolName: 'gleam',
+          constraint: config.constraints?.gleam,
+        },
+      ],
+    };
+
+    await exec('gleam deps download', execOptions);
+    const newLockFileContent = await readLocalFile(lockFileName, 'utf8');
+    /* istanbul ignore next reason: static analysis sufficient */
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.ts:60]: no

File: lib/modules/manager/gleam/artifacts.ts [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+
+  const lockFileName = getSiblingFileName(packageFileName, 'manifest.toml');
+
+  if (!lockFileName) {
+    logger.debug(`No ${lockFileName} found`);
+    return null;
+  }
+
+  const oldLockFileContent = await readLocalFile(lockFileName, 'utf8');
+  /* istanbul ignore next reason: static analysis sufficient */
```
Comment (viceice) [inline lib/modules/manager/gleam/artifacts.ts:35]: no

**Review by viceice — COMMENTED** [review]
see comments

**Comment by viceice** [comment]
why closed? 😨

**Comment by cdaringe** [comment]
Ya. I appreciate everyone's time in this.

I'm feeling a wee bit burnt out. Terse/curt PR discussion kind of exacerbates the effect for me, so I closed just to signal that I'm taking a short break.

I'd like to pick it up again in the future.

**Pull request closed without merging**